src/crimson/osd/osd_operations/pg_splitting.cc

```diff
@@ -0,0 +1,150 @@
+      if (pg->get_pgid().is_split(old_pg_num, new_pg_num, &children)) {
+        logger().debug(" {} PG has split ", pg->get_pgid());
+        logger().debug(" {} has {} children ", pg->get_pgid(), children.size());
+        std::set<Ref<PG>> split_pgs;
+        return seastar::do_with(pg, children,
+	                        std::move(next_map),
+				new_pg_num,
+				split_pgs, [this] (auto &pg,
+						   auto &children,
+						   auto &next_map,
+						   auto &new_pg_num,
+						   auto &split_pgs) {
+	  return seastar::do_for_each(children, [this, &pg, &next_map, &new_pg_num, &split_pgs] (auto& child_pg) {
+	    return shard_services.get_or_create_pg(child_pg).then([child_pg]
+		 (auto core) {
+		 logger().debug(" PG {} mapped to {}", child_pg.pgid, core);
+		 return seastar::now();
+	       }).then([this, child_pg, next_map=next_map] {
+		 logger().debug(" {} map epoch: {}", child_pg.pgid, next_map->get_epoch());
+		 return shard_services.make_pg(std::move(next_map), child_pg, true);
+	       }).then([this, &pg, &new_pg_num, &split_pgs] (Ref<PG> child_pg) {
+		 logger().debug(" Parent PG: {}", pg->get_pgid());
+		 logger().debug(" Child PG ID: {}", child_pg->get_pgid());
+
+		 const coll_t cid{child_pg->get_pgid()};
+		 unsigned split_bits = child_pg->get_pgid().get_split_bits(new_pg_num);
+		 logger().debug(" pg num is {}, m_seed is {}, split bits is {}", new_pg_num, child_pg->get_pgid().ps(), split_bits);
+		 return pg->split_colls(child_pg->get_pgid(), split_bits, child_pg->get_pgid().ps(),
+					&child_pg->get_pgpool().info, rctx.transaction).then(
+					[this, &pg, child_pg=std::move(child_pg), split_bits, &split_pgs] () {
+					logger().debug(" {} split collection done", child_pg->get_pgid());
+					pg->split_into(child_pg->get_pgid().pgid, child_pg, split_bits);
+					split_pgs.insert(child_pg);
+				      });
+		 });
+	 }).then([this, &children, &next_map, &pg, &split_pgs] {
+	   split_stats(split_pgs, children);
+	   return seastar::do_for_each(split_pgs, [this, &next_map] (auto& child_pg) {
+	       logger().debug(" {} advance map for {}", child_pg->get_pgid(), shard_services.get_map()->get_epoch());
+	       return shard_services.start_operation<PGAdvanceMap>(
+		 child_pg, shard_services, shard_services.get_map()->get_epoch(), std::move(rctx), true).second.then([this] {
+		 return seastar::now();
+	       });
+	    });
+	   });
+         });
+      }
+    } else {
+      if (new_pg_num && new_pg_num < old_pg_num) {
+	logger().debug("New PGs: {} and Old PGs: {}", new_pg_num, old_pg_num);
+	logger().debug("PG Merging is not supported yet");
+      }
+      logger().debug(" No split in {}", next_epoch);
+      return seastar::now();
+    }
+    return seastar::now();
+  });
+}
+
+void PGSplitting::split_stats(std::set<Ref<PG>> children_pgs,
+                              const std::set<spg_t> &children_pgids)
+{
+  std::vector<object_stat_sum_t> updated_stats;
+  pg->start_split_stats(children_pgids, &updated_stats);
```

Comment:
Can this function return a map <spg_t, object_stat_sum_t>? The map creation can be done after calling the inner `start_split_stats`?
This should avoiding the iterator usage below, we could iterate through the returned map